pkg/frontend/querymiddleware/codec.go

```diff
@@ -438,3 +444,57 @@ func decorateWithParamName(err error, field string) error {
 	}
 	return apierror.Newf(apierror.TypeBadData, errTmpl, field, err)
 }
+
+// marshalSampleJSON writes `[ts, "val"]`.
+// See: https://github.com/prometheus/prometheus/blob/5bd761fbfce6bf2ebc660b705f9a55477057b0c5/web/api/v1/api.go#L1640-L1652
+func marshalSampleJSON(ptr unsafe.Pointer, stm *jsoniter.Stream) {
+	smp := *((*mimirpb.Sample)(ptr))
+	stm.WriteArrayStart()
+	marshalTimestamp(smp.TimestampMs, stm)
+	stm.WriteMore()
+	marshalValue(smp.Value, stm)
+	stm.WriteArrayEnd()
+}
+
+func marshalSampleJSONIsEmpty(ptr unsafe.Pointer) bool {
+	return false
+}
+
+func marshalTimestamp(t int64, stm *jsoniter.Stream) {
```

Comment:
Would be nice if you could try to expose `marshalTimestamp()` and `marshalValue()`  in Prometheus, and move them to an util package, so that we could just import that instead of copying it. Given how tricky are these functions, I would prefer to keep them aligned with Prometheus (with the hope Prometheus maintainers don't have objections on the refactoring).

Reply:
Followed your suggestion and created a Prometheus PR to export point marshaling functions. https://github.com/prometheus/prometheus/pull/11323

I guess we could keep these copied functions at `mimirpb` and adapt mimir code later. But if you consider its better to wait until the upstream change gets merged and leave this PR on-hold, I'm ok with that.  👍

Reply:
I would move forward with this, and call Prometheus' exported functions once they're merged.